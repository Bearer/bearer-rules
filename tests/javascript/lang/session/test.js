const { createInvoker, getEnvironment } = require("../../../helper.js")
const { ruleId, ruleFile, testBase } = getEnvironment(__dirname)

describe(ruleId, () => {
  const invoke = createInvoker(ruleId, ruleFile, testBase)
  
  test("not_session_leak", ({ expect }) => {
    const testCase = "not_session_leak.js"
    expect(invoke(testCase)).toMatchSnapshot();
  })
  
  test("session_leak", ({ expect }) => {
    const testCase = "session_leak.js"
    expect(invoke(testCase)).toMatchSnapshot();
  })
  
})