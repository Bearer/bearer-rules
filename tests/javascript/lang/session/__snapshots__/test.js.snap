// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`javascript_lang_session not_session_leak 1`] = `"{}"`;

exports[`javascript_lang_session session_leak 1`] = `
"{
  "high": [
    {
      "cwe_ids": [
        "312"
      ],
      "id": "javascript_lang_session",
      "title": "Sensitive data stored in HTML local storage detected.",
      "description": "## Description\\n\\nSensitive data should not be stored in a \`localStorage\` session. This policy looks for any sensitive data stored within the localstorage.\\n\\n## Remediations\\n\\nIt's best to avoid storing sensitive data in \`localStorage\` whenever possible. To keep session data safe, use a server-based session storage solution instead.\\n\\n❌ If you do need do store data in \`localStorage\`, avoid including sensitive data:\\n\\n\`\`\`javascript\\nlocalStorage.setItem('user', email)\\n\`\`\`\\n\\n✅ Instead, use a unique identifier:\\n\\n\`\`\`javascript\\nlocalStorage.setItem('user', user.uuid)\\n\`\`\`\\n\\n## Resources\\n  - [OWASP sensitive data exposure](https://owasp.org/www-project-top-ten/2017/A3_2017-Sensitive_Data_Exposure)\\n",
      "documentation_url": "https://docs.bearer.com/reference/rules/javascript_lang_session",
      "line_number": 3,
      "full_filename": "/tmp/bearer-scan/session_leak.js",
      "filename": ".",
      "data_type": {
        "category_uuid": "14124881-6b92-4fc5-8005-ea7c1c09592e",
        "name": "Fullname"
      },
      "category_groups": [
        "PII",
        "Personal Data"
      ],
      "source": {
        "start": 1,
        "end": 1,
        "column": {
          "start": 16,
          "end": 28
        }
      },
      "sink": {
        "start": 3,
        "end": 3,
        "column": {
          "start": 1,
          "end": 35
        },
        "content": "localStorage.setItem(\\"user\\", user)"
      },
      "parent_line_number": 3,
      "snippet": "localStorage.setItem(\\"user\\", user)",
      "fingerprint": "6c0cc3faaa947ecc7d7219e846383aa3_0",
      "old_fingerprint": "09b466f85030d9662e55f4defe04fb42_0",
      "code_extract": "localStorage.setItem(\\"user\\", user);"
    }
  ]
}"
`;
